### ISSUE TITLE
Syncing with desktop Anki reverts settings changed in AnkiDroid

### ISSUE THREAD

**Issue by cameel** [ankidroid/Anki-Android#3662]
- Make sure that `New card position` is set to `Mix new cards and reviews` both in desktop Anki and in AnkiDroid and that both clients are in sync with AnkiWeb
- Go to `Settings > Reviewing` in AnkiDroid and change `New card position` to `New cards after reviews`
- Go back to deck list and sync AnkiDroid with AnkiWeb
- Sync desktop Anki with AnkiWeb
- Sync AnkiDroid with AnkiWeb
- Go back to `Settings > Reviewing`

You see that `New card position` has reverted to `Mix new cards and reviews`. The setting does not change in desktop Anki as expect it to.

I changed this setting in AnkiDroid about a week ago. Every day the setting kept its value while I was studying on the tablet (syncing AnkiDroid from time to time). In the evening I usually make changes to my decks in desktop Anki and the next day I notice the setting reverted on the tablet. This was fully reproducible during that time (and I never changed or even looked at desktop Anki preferences back then).

Today I tried changing this setting to various values on both sides to see how it behaves and sometimes it works as intended and sometimes not. For example I never managed to make it behave like this when changing setting in the other direction (i.e. from `New cards after reviews` to `Mix new cards and reviews`) or when changing it on the desktop rather than in AnkiDroid. I tried many times but the results were inconsistent. I think that's because some of client's or server's internal state has influence on this: which client synced last, whether each client thinks it needs to sync, whether it thinks its preferences have changed, etc. I can't see this state explicitly and I think that in some cases just checking it changes it (e.g. I noticed that just opening Anki desktop preferences makes it actually sync instead of just showing the `Processing...` popup). In the end I gave up and I can only say that the steps above produce this effect most of the time.

I think this affects other settings as well. For example from time to time I notice that `Next day starts at` gets reverted. I change it in AnkiDroid but after a while I notice that it has reverted again.

AnkiDroid debug info:

```
AnkiDroid Version = 2.5alpha58
Android Version = 5.1.1
ACRA UUID = aeee41c2-6522-4cb1-aaba-4cf0a8c6c06e
```

Anki:

```
Version 2.0.32
Qt 4.8.7 PyQt 4.11.4
```

**Comment by hssm** [comment]
I'm well aware of this behaviour. It's not a bug in AnkiDroid but gets brought up from time to time. It's just how the sync protocol works right now with configuration items and an annoying behaviour of the desktop client.

1 - Anki considers all collection-level settings as a single object. Thus,
2 - Settings are not *merged* on sync. They are entirely replaced with one side or the other.
3 - The desktop client marks the configuration as changed *just by opening and closing the settings window*, even if you didn't make a change.

So if you open the settings window on the desktop "just to confirm what the values are", you are, in fact, overwriting the synced settings with what you just looked at since it marks it as changed and it's newer.

I think you will find that the settings are synced correctly if you don't open this window any time before syncing both sides. (It's been a while since I looked into this so the details may be a bit off, but it's something along these lines).

**Comment by cameel** [comment]
> I think you will find that the settings are synced correctly if you don't open this window any time before syncing both sides. (It's been a while since I looked into this so the details may be a bit off, but it's something along these lines).

Unfortunately no. It happens even if the only thing I do in the desktop client is to open it, click sync button and exit.

For example now I just turned on the tablet to check, went to its settings, changed the setting to `mix cards` and synced. Then I opened desktop Anki clicked sync and closed it immediately after that. Then I synced AnkiDroid again, went to settings and I see the setting reverted.

**Comment by bretzel15** [comment]
I can confirm this bug, it happens to me for months already, even if i don't open the desktop client. Just syncing to the server with settings set to 'New cards before reviews', and after syncing on the next day, setting is back to 'mixed'.

It seems like AnkiDroid can't overwrite this setting on the server-database...

**Comment by Anthropos888** [comment]
I can confirm that a change in "New card position" on AnkiDroid will be overwritten after a sync through Anki desktop. 
I'm using the newest 2.9alpha and this bug still exists.

**Comment by Anthropos888** [comment]
This big affects custom search terms as well. When I create and save a custom search in Ankidroid and perform a sync on Anki desktop, the custom search will not be synced to Anki desktop and after the next sync on Ankidroid the search is deleted on Ankidroid.

Creating a custom search on Anki desktop will sync it as expected to Ankidroid. But the direction Ankidroid -> Desktop doesn't work.

AnkiDroid Version = 2.9alpha66

Android Version = 9

ACRA UUID = 1f29f610-893a-403f-83d5-25a26ba97fe0

**Comment by Anthropos888** [comment]
I updated Anki desktop to 2.1.11 and the above mentioned issues don't occur anymore. Everything snycs just fine in both directions. It seems that it was an issue of Anki desktop and not of Ankidroid.

**Comment by mikehardy** [comment]
well that's a happy but unexpected thing. You see anything in their changelog / diff that might explain it? You think this might be resolved and closable then?

**Comment by Anthropos888** [comment]
Before I used the last Anki desktop version before 2.1 cause I didn't want to risk any problems with scheduler v2 and AnkiDroid.

With version 2.1 up to 2.1.11 *a lot* of things got fixed in Anki desktop. So yes, I believe the newest Anki desktop version solves this issue and it can be closed.

BTW, my worries about scheduler v2 were unfounded since v1 is turned on by default in Anki 2.1.
Other good reasons to update to Anki 2.1: no more crashes in their card editor and JS is working properly now.

**Comment by mikehardy** [comment]
Okay, I'm going to mark this as closed. I've grown accustomed to finding the error on our end, so I guess I'm surprised I don't have to look through Java stack traces, but I'l take it... thanks @Anthropos888

**Comment by Anthropos888** [comment]
Sorry, but I just realized that my custom searches created in Ankidroid (alpha68) still get removed after syncing with Anki desktop (2.1.11).

Creating a custom search on Anki desktop will sync as expected to Ankidroid. But the direction Ankidroid -> Desktop doesn't work.

**Comment by mikehardy** [comment]
I think #5380 is likely duplicate of this

**Comment by mikehardy** [comment]
If anyone wants to take a look at this, it still comes up from time to time.

I think the next step is for someone with anki desktop and ankidroid installed an in use to explore how and when this fails. If you look at the database structure page on the wiki here in ankidroid you can get an idea of where settings are stored (general collection settings, deck option groups, deck configurations) and then we can determine which settings are syncing okay and which are getting clobbered, and under what conditions, so we can really understand it.

**Comment by dae** [comment]
Config syncing is solely govened by the collection modification time at the moment - whichever is newer sends the changes. There have been issues is previous releases that caused the modification time to be bumped when it shouldn't have been, so for anyone experiencing this problem, the first thing to try would be to make sure both desktop and AnkiDroid are fully up to date, and to try reproduce the issue with add-ons disabled.

**Comment by david-allison-1** [comment]
Had a quick look into things and they seem normal.

> Config syncing is solely govened by the collection modification time 

@dae  Could I confirm: this is `col.mod` in the database

**Comment by dae** [comment]
Yep
